### PR TITLE
Onboarding: Invite new users

### PR DESCRIPTION
**Pull request by dennisreimann**
First part of #5563:

- Separates the user self-registration and invite cases
- Adds invitation email for users created by the admin
- Adds invitation tokens to verify user was invited
- Adds handler action for invite links
- Refactors `UserEventHostedService`

### CODE REVIEW

File: BTCPayServer/Events/UserRegisteredEvent.cs [hunk 1]
```diff
@@ -2,14 +2,18 @@
...
+public class UserRegisteredEvent
+{
+    public ApplicationUser User { get; set; }
+    public bool Admin { get; set; }
+    public Uri RequestUri { get; set; }
+
+    public TaskCompletionSource<Uri> CallbackUrlGenerated;
+}
+
+public class UserInvitedEvent : UserRegisteredEvent
```
Comment (dennisreimann) [inline BTCPayServer/Events/UserRegisteredEvent.cs:16]: This was added to distinguish self-registration vs. invite and so that we can also at least log, who invited the new user.

File: BTCPayServer/Controllers/GreenField/GreenfieldUsersController.cs [hunk 2]
```diff
@@ -233,7 +234,9 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
                     await _settingsRepository.FirstAdminRegistered(policies, _options.UpdateUrl != null, _options.DisableRegistration, Logs);
                 }
             }
-            _eventAggregator.Publish(new UserRegisteredEvent() { RequestUri = Request.GetAbsoluteRootUri(), User = user, Admin = request.IsAdministrator is true });
+
+            var currentUser = await _userManager.GetUserAsync(User);
+            _eventAggregator.Publish(new UserInvitedEvent { RequestUri = Request.GetAbsoluteRootUri(), User = user, Admin = isNewAdmin, InvitedByUser = currentUser });
```
Comment (dennisreimann) [inline BTCPayServer/Controllers/GreenField/GreenfieldUsersController.cs:239]: I opted to change this to invite, because the user does not self-register and no password is set. It will trigger an email and the user can setup (email confirmation, set password) via the new invite handler action.

File: BTCPayServer/Controllers/UIAccountController.cs [hunk 3]
```diff
@@ -784,6 +783,33 @@ public async Task<IActionResult> SetPassword(SetPasswordViewModel model)
             return View(model);
         }
 
+        [AllowAnonymous]
+        [HttpGet("/invite/{userId}/{code}")]
+        public async Task<IActionResult> AcceptInvite(string userId, string code)
```
Comment (dennisreimann) [inline BTCPayServer/Controllers/UIAccountController.cs:788]: One action to handle different cases (email confirmation, set password) uniformly.

File: BTCPayServer/Extensions/EmailSenderExtensions.cs [hunk 4]
```diff
@@ -29,11 +29,17 @@ public static void SendApprovalConfirmation(this IEmailSender emailSender, Mailb
                 $"Your account has been approved and you can now <a href='{HtmlEncoder.Default.Encode(link)}'>login here</a>");
         }
 
-        public static void SendSetPasswordConfirmation(this IEmailSender emailSender, MailboxAddress address, string link, bool newPassword)
+        public static void SendSetPassword(this IEmailSender emailSender, MailboxAddress address, string link, bool hasPassword)
```
Comment (dennisreimann) [inline BTCPayServer/Extensions/EmailSenderExtensions.cs:32]: Changed the name and argument to be a bit clearer.

**Comment by dennisreimann** [comment]
Will add tests, but wanted to publish the current state so that we can test and discuss the open points from #5563, @dstrukt @pavlenex.

We could also split the issue between two PRs and the user invite on store-level could be tackled in a follow-up.

File: BTCPayServer/Extensions/EmailSenderExtensions.cs [hunk 5]
```diff
@@ -19,21 +19,26 @@ private static string CallToAction(string actionName, string actionLink)
...
 
         public static void SendApprovalConfirmation(this IEmailSender emailSender, MailboxAddress address, string link)
         {
-            emailSender.SendEmail(address, "Your account has been approved",
+            emailSender.SendEmail(address, "BTCPay Server: Your account has been approved",
                 $"Your account has been approved and you can now <a href='{HtmlEncoder.Default.Encode(link)}'>login here</a>");
         }
 
-        public static void SendSetPasswordConfirmation(this IEmailSender emailSender, MailboxAddress address, string link, bool newPassword)
+        public static void SendSetPassword(this IEmailSender emailSender, MailboxAddress address, string link)
```
Comment (dennisreimann) [inline BTCPayServer/Extensions/EmailSenderExtensions.cs:32]: This isn't used for initial password setting anymore, because the invitation action takes care of that.

**Comment by dennisreimann** [comment]
I think it makes sense to keep the scope of this PR to what it is right now. I can do the role updates and invite on the store level in another PR building on this. 

So feel free to review @pavlenex and @dstrukt. Main points from the user/admin perspective:

- Separation of user self-registration and invite cases
- Invitation email for users created by the admin
- Accounts created by an admin are now auto-approved
- Admins receive an email if a new account requires approval

**Comment by pavlenex** [comment]
cACK. 

Is there anything specific that needs to be user-tested here? In general what is proposed sounds good to me.

**Comment by pavlenex** [comment]
Tested:

- Invitation email for users created by the admin  ✅ 
- Accounts created by an admin are now auto-approved  ✅ 
- Admins receive an email if a new account requires approval 🚧 (still testing)
   - @dennisreimann do all admins receive email? 

### Bugs encountered 🐛 
#### When changing user to admin role or removing and admin role, we spit out an error,`Error updating the user` but the changes still apply successfully. See video. May be unrelated to this PR.
https://github.com/btcpayserver/btcpayserver/assets/36959754/74dd4407-5155-450e-a09b-759e9284127e

**Comment by dennisreimann** [comment]
> do all admins receive email once user is pending approval?

Yes. All admins should receive th email and notification about the new user requiring approval.

> BUG 1 When changing user to admin role or removing and admin role, we spit out an error,Error updating the user but the changes still apply successfully.

Fixed.

> BUG 2 When email verification is enabled, and user is added manually by admin, email verification email is never sent, but user receives account creation email.

This is now also fixed. There is no separate "Confirm your email address" email, as we can make this part of the process when accepting the invitation — so everything is handled by the invitation email the user got. However, the problem was that the email confirmation in that action did not work. Now it does — thanks for spotting this error! :)

**Comment by dennisreimann** [comment]
Updated as discussed with @pavlenex: Send notification to admins only if the user does ...

- either not require email confirmation
- or when they confirmed their email address
 
Rationale: We want to inform admins only about qualified users and not annoy them with bot registrations in case email confirmation is required.

**Comment by pavlenex** [comment]
@dennisreimann Found one minor thing. These are my server settings
Screenshot 2024-02-09 at 12 35 22

I invite the user and Admin gets notification that user is waiting approval, but I think all users invited by admin are approved, so this is probabaly notification code remaining from one of previous changes, can you check please? Notification says user waiting approval, but the user is already approved basically when you go to that user.

**Review by dstrukt — APPROVED** [review]
Finally got this working on my local machine, basically nuked and started fresh.

Tested the 3 cases @pavlenex mentioned previously that were supposed fixed - looks like they work properly for me.

I'm sure there's more testing and edge cases that can be done, but maybe kukks or nico will find something!

File: BTCPayServer/Events/UserRegisteredEvent.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dennisreimann) [reply BTCPayServer/Events/UserRegisteredEvent.cs:16]: Imho this variant is more explicit and feels better than cluttering up the more general event with a flag plus extra property. I've [added the subclass to the list of events that are subscribed](https://github.com/btcpayserver/btcpayserver/pull/5714/files#diff-28cf1ff2a014e577e74f91b7437d1e360b2968aafe2e835941a313c1e5e9762cR31) as well.

File: BTCPayServer/Security/IdentityBuilderExtension.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+using Microsoft.AspNetCore.Identity;
+
+namespace BTCPayServer.Security;
+
+public static class IdentityBuilderExtension
+{
+    public static IdentityBuilder AddInvitationTokenProvider(this IdentityBuilder builder)
+    {
+        var userType = builder.UserType;
```
[the remark this answers is not in the document]
Comment (dennisreimann) [reply BTCPayServer/Security/IdentityBuilderExtension.cs]: Done in 3fe482a.

File: BTCPayServer/Events/UserRegisteredEvent.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dennisreimann) [reply BTCPayServer/Events/UserRegisteredEvent.cs]: Updated in a823aff.
